keps/sig-instrumentation/3498-extending-stability/README.md

```diff
@@ -0,0 +1,179 @@
+
+
+### Risks and Mitigations
+
+The primary risk is that these changes break our existing (and working) metrics infrastructure. The mitigation should straightfoward, i.e. rollback the changes to the metrics framework. 
+
+## Design Details
+
+Our plan is to add functionality to our static analysis framework which is hosted in the main `k8s/k8s` repo, under `test/instrumentation`. Specifically, we will need to support:
+
+- parsing variables
+- multi-line strings
+- evaluating buckets
+- buckets which are defined via variables and consts
+- evaluation of simple consts
+- evaluation of simple variables
+
+We will not attempt to parse metrics which:
+
+- are constructed dynamically, i.e. through function calls which use function arguments as parameters in metric definitions, since some of those cannot be resolved until runtime. 
+- are constructed using custom prometheus collectors, for the same reasons as above. 
+
+As an aside, much of this work has already been done, but is stashed in a local repo. 
+
+### Test Plan
+
+We have static analysis testing for stable metrics, we will extend our test coverage 
+to include metrics which are `ALPHA` and `BETA` while ignoring `INTERNAL` metrics.
+
+
+### Graduation Criteria
+
+
+#### Alpha
+
+- Kubernetes metrics framework will be enhanced to support additional stability classes
+- The static analysis pipeline of the metrics framework will be enhanced to understand how to parse more things (these are listed above)
+- All instances of `Alpha` metrics will be converted to `Internal`
+
+#### Beta
+
+- Kubernetes metrics framework will be enhanced to support marking `Alpha` and `Beta` metrics with a date. The semantics of this are yet to be determined. This date will be used to statically determine whether or not that metric should be decrepated automatically or promoted.
+- Kubernetes metrics framework will be enhanced with a script to auto-deprecate metrics which have passed their window of existence as an `Alpha` or `Beta` metric
+- It is at this point, we will determine the longevity rules for `Alpha` and `Beta` metrics
+- The `beta` stage for this framework will be a few releases. During this time, we will evaluate the utility and the ergonomics of the framework, making adjustments as necessary
+
+#### GA
+
+
+## Production Readiness Review Questionnaire
+
+During the `alpha` stage of this KEP, we will not be making any user facing changes, except marking metrics as `Internal` which were previously `Alpha`. The stability guarantees of `Internal` metrics is the same as `Alpha` currently and therefore there will not be any changes to what users can expect from the metrics they are using. 
+
+### Feature Enablement and Rollback
+
+We can revert our changes if it breaks the metrics framework. But we will be adding testing coverage as we enhance the framework, so it is unlikely that this will need to occur.
```

Comment:
Would it even be possible to somehow hide them behind feature gate? If so, that would be preferrable...

If not, I agree it's not a feature per-se, rather internal, semi-static enablement, and it would make sense to make it explicit here.

Reply:
No, I mean it's adding enum values to the framework.